File: pkg/retrieval/retrieval.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

// Package retrieval defines transport-neutral knowledge retrieval contracts.
package retrieval

import (
	"context"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"github.com/phrocker/shoal-oss/pkg/document"
	"github.com/phrocker/shoal-oss/pkg/graph"
	"github.com/phrocker/shoal-oss/pkg/shoal"
)

const (
	DefaultTopK         uint32 = 20
	MaxTopK             uint32 = 1000
	MaxQueryBytes              = 16 * 1024
	MaxScopeIDs                = 10_000
	MaxModes                   = 4
	MaxExplanationBytes        = 16 * 1024

	// MaxIdempotencyTokenBytes is reserved for future additive generic
	// mutation APIs. Retrieval requests do not carry such a token.
	MaxIdempotencyTokenBytes = 128
)

// Mode selects a retrieval strategy. Multiple modes request a hybrid plan.
type Mode string

const (
	ModeLexical Mode = "lexical"
	ModeVector  Mode = "vector"
	ModeTree    Mode = "tree"
	ModeGraph   Mode = "graph"
)

// Scope bounds retrieval to known documents or graph nodes. IDs are ORed
// within each list. When both lists are nonempty, candidates must satisfy both
// dimensions through a canonical document/graph association.
type Scope struct {
	DocumentIDs []shoal.ID
	NodeIDs     []shoal.ID
}

// Request describes one coarse knowledge retrieval operation.
type Request struct {
	Text    string
	TopK    uint32
	Modes   []Mode
	Scope   Scope
	AsOf    time.Time
	Explain bool
}

// Normalize returns an independently owned, canonical request and never
// mutates caller slices. It applies public defaults, removes duplicate modes
// and scope IDs while preserving first occurrence, and normalizes AsOf to UTC.
func (r Request) Normalize() (Request, error) {
	normalized := r
	if normalized.TopK == 0 {
		normalized.TopK = DefaultTopK
	}
	if len(r.Modes) == 0 {
		normalized.Modes = []Mode{ModeLexical}
	} else {
		normalized.Modes = deduplicateModes(r.Modes)
	}
	normalized.Scope = Scope{
		DocumentIDs: deduplicateIDs(r.Scope.DocumentIDs),
		NodeIDs:     deduplicateIDs(r.Scope.NodeIDs),
	}
	if !r.AsOf.IsZero() {
		normalized.AsOf = r.AsOf.UTC()
	}
	if err := normalized.validateNormalized(); err != nil {
		return Request{}, err
	}
	return normalized, nil
}

// Validate checks transport-independent request invariants after public
// normalization.
func (r Request) Validate() error {
	_, err := r.Normalize()
	return err
}

func (r Request) validateNormalized() error {
	if !utf8.ValidString(r.Text) {
		return shoal.NewError(shoal.ErrorInvalidArgument, "retrieval text must be valid UTF-8")
	}
	if strings.TrimSpace(r.Text) == "" {
		return shoal.NewError(shoal.ErrorInvalidArgument, "retrieval text is required")
	}
	if len(r.Text) > MaxQueryBytes {
		return shoal.NewError(
			shoal.ErrorInvalidArgument, "retrieval text exceeds the public byte bound")
	}
	if r.TopK == 0 || r.TopK > MaxTopK {
		return shoal.NewError(
			shoal.ErrorInvalidArgument, "retrieval top_k is outside the public bound")
	}
	if len(r.Modes) == 0 || len(r.Modes) > MaxModes {
		return shoal.NewError(
			shoal.ErrorInvalidArgument, "retrieval mode count is outside the public bound")
	}
	for _, mode := range r.Modes {
		if !validMode(mode) {
			return shoal.NewError(shoal.ErrorInvalidArgument, "unknown retrieval mode")
		}
	}
	if len(r.Scope.DocumentIDs)+len(r.Scope.NodeIDs) > MaxScopeIDs {
		return shoal.NewError(
			shoal.ErrorInvalidArgument, "retrieval scope has too many IDs")
	}
	for _, id := range r.Scope.DocumentIDs {
		if err := shoal.ValidateRequiredID("retrieval document ID", id); err != nil {
			return err
		}
	}
	for _, id := range r.Scope.NodeIDs {
		if err := shoal.ValidateRequiredID("retrieval node ID", id); err != nil {
			return err
		}
	}
	if !r.AsOf.IsZero() {
		year := r.AsOf.Year()
		if year < 1 || year > 9999 {
			return shoal.NewError(
				shoal.ErrorInvalidArgument, "retrieval as_of is outside the supported range")
		}
	}
	return nil
}

// ValidateSeedPlan rejects tree or graph shapes that have no bounded seed
// source under the fixed lexical, tree, then graph planner. The boolean states
// whether document-to-graph association seeds are implemented by the target.
func (r Request) ValidateSeedPlan(documentAssociationSeeds bool) error {
	normalized, err := r.Normalize()
	if err != nil {
		return err
	}
	hasLexical := normalized.HasMode(ModeLexical)
	hasVector := normalized.HasMode(ModeVector)
	hasTree := normalized.HasMode(ModeTree)
	hasGraph := normalized.HasMode(ModeGraph)

	treeSeeded := len(normalized.Scope.DocumentIDs) > 0 || hasLexical || hasVector
	if hasTree && !treeSeeded {
		return shoal.NewError(
			shoal.ErrorUnavailable,
			"tree retrieval requires document scope or earlier lexical/vector seeds",
		)
	}
	graphSeeded := len(normalized.Scope.NodeIDs) > 0 ||
		(documentAssociationSeeds && len(normalized.Scope.DocumentIDs) > 0) ||
		hasLexical || hasVector || (hasTree && treeSeeded)
	if hasGraph && !graphSeeded {
		return shoal.NewError(
			shoal.ErrorUnavailable,
			"graph retrieval requires node scope, document associations, or earlier seeds",
		)
	}
	return nil
}

// HasMode reports whether a request contains a mode.
func (r Request) HasMode(mode Mode) bool {
	for _, candidate := range r.Modes {
		if candidate == mode {
			return true
		}
	}
	return false
}

// Evidence ties a result to immutable source and, when applicable, the graph
// path used to reach it.
type Evidence struct {
	Citation document.Citation
	Quote    string
	Path     graph.Path
	Score    shoal.Score
}

// Validate checks transport-neutral evidence structure. Quote equality is a
// document-context check performed against canonical retained source bytes.
func (e Evidence) Validate() error {
	if err := e.Citation.Validate(); err != nil {
		return err
	}
	if err := shoal.ValidateFiniteScore("evidence score", e.Score); err != nil {
		return err
	}
	if pathPresent(e.Path) {
		if err := e.Path.Validate(); err != nil {
			return err
		}
	}
	return nil
}

// Explanation describes why a result was selected without exposing an
// execution engine or storage plan.
type Explanation struct {
	Modes   []Mode
	Summary string
	Scores  map[string]shoal.Score
}

// Result is one ranked, evidence-addressable retrieval result.
type Result struct {
	ID          shoal.ID
	Score       shoal.Score
	Evidence    []Evidence
	Explanation *Explanation
}

// Response is the complete, all-or-error result of one retrieval request.
type Response struct {
	RequestID shoal.ID
	Results   []Result
}

// ValidateFor checks result bounds, uniqueness, finite scores, and normative
// deterministic result/evidence ordering for a normalized request.
func (r Response) ValidateFor(request Request) error {
	normalized, err := request.Normalize()
	if err != nil {
		return err
	}
	if err := shoal.ValidateOptionalID("retrieval request ID", r.RequestID); err != nil {
		return err
	}
	if uint32(len(r.Results)) > normalized.TopK {
		return shoal.NewError(
			shoal.ErrorInvalidArgument, "retrieval response exceeds normalized top_k")
	}
	seen := make(map[shoal.ID]struct{}, len(r.Results))
	for index, result := range r.Results {
		if err := shoal.ValidateRequiredID("retrieval result ID", result.ID); err != nil {
			return err
		}
		if _, duplicate := seen[result.ID]; duplicate {
			return shoal.NewError(
				shoal.ErrorInvalidArgument, "retrieval response has duplicate result IDs")
		}
		seen[result.ID] = struct{}{}
		if err := shoal.ValidateFiniteScore("retrieval result score", result.Score); err != nil {
			return err
		}
		if index > 0 && CompareResult(r.Results[index-1], result) > 0 {
			return shoal.NewError(
				shoal.ErrorInvalidArgument, "retrieval results are not deterministically ordered")
		}
		for evidenceIndex, evidence := range result.Evidence {
			if err := evidence.Validate(); err != nil {
				return err
			}
			if evidenceIndex > 0 {
				compared := CompareEvidence(result.Evidence[evidenceIndex-1], evidence)
				if compared > 0 {
					return shoal.NewError(
						shoal.ErrorInvalidArgument,
						"retrieval evidence is not deterministically ordered",
					)
				}
				if compared == 0 {
					return shoal.NewError(
						shoal.ErrorInvalidArgument,
						"retrieval evidence has a duplicate ordering key",
					)
				}
			}
		}
		if result.Explanation != nil {
			if err := validateExplanation(*result.Explanation); err != nil {
				return err
			}
		}
	}
	return nil
}

// CompareResult returns a negative value when left sorts before right under
// score-descending, raw-result-ID-ascending order.
func CompareResult(left, right Result) int {
	switch {
	case left.Score > right.Score:
		return -1
	case left.Score < right.Score:
		return 1
	default:
		return shoal.CompareID(left.ID, right.ID)
	}
}

// CompareEvidence compares the public evidence ordering key: score descending,
// citation tuple, then directed path tuple. A response must not contain two
// evidence values with the same ordering key.
func CompareEvidence(left, right Evidence) int {
	switch {
	case left.Score > right.Score:
		return -1
	case left.Score < right.Score:
		return 1
	}
	if compared := compareCitation(left.Citation, right.Citation); compared != 0 {
		return compared
	}
	return comparePath(left.Path, right.Path)
}

// SortResults orders results and each result's evidence by the public total
// orders. It mutates the supplied response slices but not nested values.
func SortResults(response *Response) {
	if response == nil {
		return
	}
	sort.SliceStable(response.Results, func(i, j int) bool {
		return CompareResult(response.Results[i], response.Results[j]) < 0
	})
	for index := range response.Results {
		sort.SliceStable(response.Results[index].Evidence, func(i, j int) bool {
			return CompareEvidence(
				response.Results[index].Evidence[i],
				response.Results[index].Evidence[j],
			) < 0
		})
	}
}

// Retriever is implemented by embedded and remote Shoal clients.
type Retriever interface {
	Retrieve(context.Context, Request) (Response, error)
}

func deduplicateModes(modes []Mode) []Mode {
	result := make([]Mode, 0, len(modes))
	seen := make(map[Mode]struct{}, len(modes))
	for _, mode := range modes {
		if _, duplicate := seen[mode]; duplicate {
			continue
		}
		seen[mode] = struct{}{}
		result = append(result, mode)
	}
	return result
}

func deduplicateIDs(ids []shoal.ID) []shoal.ID {
	if len(ids) == 0 {
		return nil
	}
	result := make([]shoal.ID, 0, len(ids))
	seen := make(map[shoal.ID]struct{}, len(ids))
	for _, id := range ids {
		if _, duplicate := seen[id]; duplicate {
			continue
		}
		seen[id] = struct{}{}
		result = append(result, id)
	}
	return result
}

func validMode(mode Mode) bool {
	switch mode {
	case ModeLexical, ModeVector, ModeTree, ModeGraph:
		return true
	default:
		return false
	}
}

func validateExplanation(explanation Explanation) error {
	if len(explanation.Summary) > MaxExplanationBytes {
		return shoal.NewError(
			shoal.ErrorInvalidArgument, "explanation summary exceeds the public byte bound")
	}
	for _, mode := range explanation.Modes {
		if !validMode(mode) {
			return shoal.NewError(
				shoal.ErrorInvalidArgument, "explanation contains an unknown mode")
		}
	}
	for name, score := range explanation.Scores {
		if err := shoal.ValidateSemanticString("explanation score name", name); err != nil {
			return err
		}
		if err := shoal.ValidateFiniteScore("explanation score", score); err != nil {
			return err
		}
	}
	return nil
}

func compareCitation(left, right document.Citation) int {
	for _, pair := range [][2]shoal.ID{
		{left.DocumentID, right.DocumentID},
		{left.RevisionID, right.RevisionID},
		{left.SectionID, right.SectionID},
		{left.SpanID, right.SpanID},
	} {
		if compared := shoal.CompareID(pair[0], pair[1]); compared != 0 {
			return compared
		}
	}
	switch {
	case left.Range.Start.Offset < right.Range.Start.Offset:
		return -1
	case left.Range.Start.Offset > right.Range.Start.Offset:
		return 1
	case left.Range.End.Offset < right.Range.End.Offset:
		return -1
	case left.Range.End.Offset > right.Range.End.Offset:
		return 1
	default:
		return 0
	}
}

func comparePath(left, right graph.Path) int {
	for index := 0; index < len(left.Nodes) && index < len(right.Nodes); index++ {
		if compared := shoal.CompareID(left.Nodes[index].ID, right.Nodes[index].ID); compared != 0 {
			return compared
		}
	}
	if len(left.Nodes) < len(right.Nodes) {
		return -1
	}
	if len(left.Nodes) > len(right.Nodes) {
		return 1
	}
	for index := 0; index < len(left.Edges) && index < len(right.Edges); index++ {
		if compared := shoal.CompareID(left.Edges[index].ID, right.Edges[index].ID); compared != 0 {
			return compared
		}
	}
	switch {
	case len(left.Edges) < len(right.Edges):
		return -1
	case len(left.Edges) > len(right.Edges):
		return 1
	default:
		return 0
	}
}

func pathPresent(path graph.Path) bool {
	return len(path.Nodes) > 0 || len(path.Edges) > 0
}

File: python/src/shoal/retrieval.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements. See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership. The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License. You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied. See the License for the
# specific language governing permissions and limitations
# under the License.

"""Transport-neutral knowledge retrieval contracts."""

from __future__ import annotations

from dataclasses import field
from datetime import datetime, timezone
from types import MappingProxyType
from typing import List, Mapping, Optional, Protocol, Tuple, Union, runtime_checkable

from ._compat import StrEnum, frozen_dataclass
from .document import Citation
from .errors import ErrorCode, new_error
from .graph import Path
from .types import ID, Score


class Mode(StrEnum):
    """A retrieval strategy; multiple modes request a hybrid plan."""

    LEXICAL = "lexical"
    VECTOR = "vector"
    TREE = "tree"
    GRAPH = "graph"


@frozen_dataclass
class Scope:
    """Bounds retrieval to known documents or graph nodes."""

    document_ids: tuple[ID, ...] = ()
    node_ids: tuple[ID, ...] = ()

    def __post_init__(self) -> None:
        object.__setattr__(self, "document_ids", tuple(self.document_ids))
        object.__setattr__(self, "node_ids", tuple(self.node_ids))

    def validate(self) -> None:
        """Validate identifiers for transport as protobuf strings."""

        for name, values in (
            ("document ID", self.document_ids),
            ("node ID", self.node_ids),
        ):
            for value in values:
                if not _is_wire_string(value):
                    raise new_error(
                        ErrorCode.INVALID_ARGUMENT,
                        f"retrieval {name} must be a valid UTF-8 string",
                    )


@frozen_dataclass
class Request:
    """One coarse knowledge retrieval operation."""

    text: str = ""
    top_k: int = 0
    modes: Tuple[Union[Mode, str], ...] = ()
    scope: Scope = field(default_factory=Scope)
    as_of: Optional[datetime] = None
    explain: bool = False

    def __post_init__(self) -> None:
        normalized_modes: List[Union[Mode, str]] = []
        for mode in self.modes:
            try:
                normalized_modes.append(Mode(mode))
            except (TypeError, ValueError):
                normalized_modes.append(mode)
        object.__setattr__(self, "modes", tuple(normalized_modes))

    def validate(self) -> None:
        """Validate transport-independent request invariants."""

        if not _is_wire_string(self.text):
            raise new_error(
                ErrorCode.INVALID_ARGUMENT,
                "retrieval text must be a valid UTF-8 string",
            )
        if not self.text.strip():
            raise new_error(
                ErrorCode.INVALID_ARGUMENT,
                "retrieval text is required",
            )
        if (
            isinstance(self.top_k, bool)
            or not isinstance(self.top_k, int)
            or self.top_k < 0
            or self.top_k > 2**32 - 1
        ):
            raise new_error(
                ErrorCode.INVALID_ARGUMENT,
                "top_k must be an unsigned 32-bit integer",
            )
        for mode in self.modes:
            try:
                Mode(mode)
            except (TypeError, ValueError):
                raise new_error(
                    ErrorCode.INVALID_ARGUMENT,
                    "unknown retrieval mode",
                ) from None
        if not isinstance(self.scope, Scope):
            raise new_error(
                ErrorCode.INVALID_ARGUMENT,
                "scope must be a Scope",
            )
        self.scope.validate()
        if self.as_of is not None:
            if not isinstance(self.as_of, datetime):
                raise new_error(
                    ErrorCode.INVALID_ARGUMENT,
                    "as_of must be a datetime",
                )
            if self.as_of.utcoffset() is None:
                raise new_error(
                    ErrorCode.INVALID_ARGUMENT,
                    "as_of must be timezone-aware",
                )
            try:
                utc_as_of = self.as_of.astimezone(timezone.utc)
            except (OverflowError, ValueError):
                raise new_error(
                    ErrorCode.INVALID_ARGUMENT,
                    "as_of is outside the supported UTC range",
                ) from None
            if utc_as_of.year < 1 or utc_as_of.year > 9999:
                raise new_error(
                    ErrorCode.INVALID_ARGUMENT,
                    "as_of is outside the supported UTC range",
                )


def _is_wire_string(value: object) -> bool:
    if not isinstance(value, str):
        return False
    try:
        value.encode("utf-8")
    except UnicodeEncodeError:
        return False
    return True


@frozen_dataclass
class Evidence:
    """Immutable source evidence and an optional graph path."""

    citation: Citation = field(default_factory=Citation)
    quote: str = ""
    path: Path = field(default_factory=Path)
    score: Score = 0.0


@frozen_dataclass
class Explanation:
    """Why a result was selected, without an execution or storage plan."""

    modes: tuple[Mode, ...] = ()
    summary: str = ""
    scores: Mapping[str, Score] = field(
        default_factory=lambda: MappingProxyType({})
    )

    def __post_init__(self) -> None:
        object.__setattr__(self, "modes", tuple(self.modes))
        object.__setattr__(
            self,
            "scores",
            MappingProxyType(dict(self.scores)),
        )


@frozen_dataclass
class Result:
    """One ranked, evidence-addressable retrieval result."""

    id: ID = ""
    score: Score = 0.0
    evidence: tuple[Evidence, ...] = ()
    explanation: Optional[Explanation] = None

    def __post_init__(self) -> None:
        object.__setattr__(self, "evidence", tuple(self.evidence))


@frozen_dataclass
class Response:
    """The complete result of one retrieval request."""

    request_id: ID = ""
    results: tuple[Result, ...] = ()

    def __post_init__(self) -> None:
        object.__setattr__(self, "results", tuple(self.results))


@runtime_checkable
class Retriever(Protocol):
    """The transport-neutral retrieval client contract."""

    def retrieve(self, request: Request) -> Response:
        """Retrieve ranked knowledge for ``request``."""

        ...


@runtime_checkable
class NativeTransport(Protocol):
    """The boundary a future native transport must implement."""

    def retrieve(self, request: Request) -> Response:
        """Perform a native retrieval operation."""

        ...


@runtime_checkable
class RemoteTransport(Protocol):
    """The boundary a future remote transport must implement."""

    def retrieve(self, request: Request) -> Response:
        """Perform a remote retrieval operation."""

        ...


@frozen_dataclass
class NativeRetriever:
    """Validating adapter for a native retrieval transport."""

    transport: NativeTransport

    def retrieve(self, request: Request) -> Response:
        request.validate()
        return self.transport.retrieve(request)


@frozen_dataclass
class RemoteRetriever:
    """Validating adapter for a remote retrieval transport."""

    transport: RemoteTransport

    def retrieve(self, request: Request) -> Response:
        request.validate()
        return self.transport.retrieve(request)
